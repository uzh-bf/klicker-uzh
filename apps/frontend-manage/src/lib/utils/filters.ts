/* eslint-disable no-unused-vars */
// TODO: typing

import dayjs from 'dayjs'
import _every from 'lodash/every'
// import Fuse from 'fuse.js'
import * as JsSearch from 'js-search'

const indices = {}

/* const INDEX_CONFIGS = {
  questions: {
    location: 0,
    distance: 100,
    threshold: 0.5,
    maxPatternLength: 32,
    minMatchCharLength: 1,
    tokenize: true,
    matchAllTokens: true,
    includeScore: true,
    includeMatches: true,
    // TODO: add description to search keys?
    keys: ['title', 'tags.name', 'type'],
  },
} */

export function buildIndex(
  name: string,
  items: any[],
  searchIndices: any[]
): any[] {
  // if an index already exists, return it

  if (indices[name]) {
    return indices[name]
  }

  // build a new js-search index
  const search = new JsSearch.Search('id')

  // use the TF-IDF strategy
  search.searchIndex = new JsSearch.TfIdfSearchIndex()

  // look for all substrings, not only prefixed
  search.indexStrategy = new JsSearch.AllSubstringsIndexStrategy()

  if (name !== 'users') {
    // impedes the user search, must it be included for the other types?
    search.tokenizer = new JsSearch.StopWordsTokenizer(
      new JsSearch.SimpleTokenizer()
    )
  }

  // index by title, type, creation date and the description of the first version
  searchIndices.forEach((index): void => search.addIndex(index))

  // build the index based on the items
  search.addDocuments(items)

  // store the index and return it (singleton)
  indices[name] = search
  return search
}

export function filterQuestions(
  questions: any[],
  filters: any,
  index: any
): any[] {
  let results = [...questions]

  // if a title (query) was given, search the index with it
  if (index && filters.title) {
    results = index.search(filters.title)
  }

  // only reduce the shown questions to the non-archived ones, if the archive filter is not active
  if (!filters.archive) {
    results = results.filter(
      ({ isArchived }): boolean =>
        (typeof isArchived === 'undefined' && !filters.archive) ||
        isArchived === false
    )
  }

  // if either type or tags were selected, filter the results
  if (filters.type || filters.tags) {
    results = results.filter(({ type, tags }): boolean => {
      // compare the type selected and the type of each question
      if (filters.type && type !== filters.type) {
        return false
      }

      // compare the tags selected and check whether the question fulfills all of them
      if (
        filters.tags &&
        !_every(filters.tags, (tag): boolean =>
          tags.map((t): string => t.name).includes(tag)
        )
      ) {
        return false
      }

      return true
    })
  }

  return results
}

// both used for sessions and users
export function filterByTitle(objects: any[], filters: any, index: any): any[] {
  let results = objects

  if (filters.title) {
    results = index.search(filters.title)
  }

  return results
}

export function subtractDates(date1: any, date2: any): any {
  return date1 - date2
}

export function sortQuestions(questions: any[], sort: any): any[] {
  const factor = sort.asc ? 1 : -1

  if (sort.by === 'TITLE') {
    return questions.sort(
      (a, b): number => factor * a.title.localeCompare(b.title)
    )
  }

  if (sort.by === 'TYPE') {
    return questions.sort(
      (a, b): number => factor * a.type.localeCompare(b.type)
    )
  }

  if (sort.by === 'CREATED') {
    return questions.sort(
      (a, b): number =>
        factor * subtractDates(dayjs(a.createdAt), dayjs(b.createdAt))
    )
  }

  if (sort.by === 'USED') {
    return questions.sort((a, b): number => {
      if (a.instances.length === 0 || b.instances.length === 0) {
        if (a.instances.length === 0 && b.instances.length === 0) {
          return 0
        }

        return factor * 1
      }

      // compare the dates of the latest created instances
      // this allows us to sort by "last usage"
      return (
        factor *
        subtractDates(
          dayjs(a.instances[a.instances.length - 1].createdAt),
          dayjs(b.instances[b.instances.length - 1].createdAt)
        )
      )
    })
  }

  return questions
}

export function processItems(items, filters, sort, index): any[] {
  let processed = items

  if (filters) {
    processed = filterQuestions(processed, filters, index)
  }

  if (sort) {
    processed = sortQuestions(processed, sort)
  }

  return processed
}
