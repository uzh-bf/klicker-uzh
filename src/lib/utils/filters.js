// @flow
/* eslint-disable no-unused-vars */

import _every from 'lodash/every'

import type { QuestionType } from '../../queries/queries'

type QuestionFilters = {
  tags: Array<string>,
  title: ?string,
  type: ?string,
}

function filterQuestions(questions: QuestionType[], filters: QuestionFilters) {
  return questions.filter((question) => {
    if (filters.title && !question.title.includes(filters.title)) {
      return false
    }
    if (
      filters.tags &&
      !_every(filters.tags, tag => question.tags.map(t => t.name).includes(tag))
    ) {
      return false
    }
    if (filters.type && question.type !== filters.type) {
      return false
    }
    return true
  })
}

function filterSessions(sessions: any, filters: any) {
  return sessions
}

export { filterQuestions, filterSessions }
export type { QuestionFilters }
