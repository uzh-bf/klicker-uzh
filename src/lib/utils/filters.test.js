import { filterQuestions } from './filters'

const questions = [
  {
    id: '1',
    tags: ['Tag 1', 'Tag 2', 'Tag 3'],
    title: 'Question 1',
    type: 'SC',
  },
  {
    id: '2',
    tags: ['Tag 1', 'Tag 4', 'Tag 5'],
    title: 'Question 2',
    type: 'MC',
  },
  {
    id: '3',
    tags: ['Tag 1', 'Tag 6', 'Tag 7'],
    title: 'Question 3',
    type: 'FREE',
  },
  {
    id: '4',
    tags: ['Tag 1', 'Tag 4', 'Tag 5'],
    title: 'Question 4',
    type: 'SC',
  },
]

expect.addSnapshotSerializer({
  print: val => `Question #${val.id}
  - Title: ${val.title}
  - Tags: ${val.tags}
  - Type: ${val.type}`,
  test: val => val.id && val.tags && val.title && val.type,
})

describe('filterQuestions', () => {
  it('returns all on empty filter', () => {
    expect(filterQuestions(questions, {})).toMatchSnapshot()
  })
  it('filters by title', () => {
    expect(filterQuestions(questions, { title: 'tion 1' })).toMatchSnapshot()
    expect(filterQuestions(questions, { title: 'nonexistent' })).toMatchSnapshot()
  })
  it('filters by tags', () => {
    expect(filterQuestions(questions, { tags: ['Tag 1'] })).toMatchSnapshot()
    expect(filterQuestions(questions, { tags: ['Tag 6'] })).toMatchSnapshot()
    expect(filterQuestions(questions, { tags: ['Tag 1', 'Tag 6'] }))
    expect(filterQuestions(questions, { tags: ['nonexistent'] })).toMatchSnapshot()
  })
  it('filters by type', () => {
    expect(filterQuestions(questions, { type: 'SC' })).toMatchSnapshot()
    expect(filterQuestions(questions, { type: 'nonexistent' })).toMatchSnapshot()
  })
})
