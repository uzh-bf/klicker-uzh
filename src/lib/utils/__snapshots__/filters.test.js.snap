// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`filterQuestions filters by tags 1`] = `
Array [
  Object {
    "id": "1",
    "tags": Array [
      "Tag 1",
      "Tag 2",
      "Tag 3",
    ],
    "title": "Question 1",
    "type": "SC",
  },
  Object {
    "id": "2",
    "tags": Array [
      "Tag 1",
      "Tag 4",
      "Tag 5",
    ],
    "title": "Question 2",
    "type": "MC",
  },
  Object {
    "id": "3",
    "tags": Array [
      "Tag 1",
      "Tag 6",
      "Tag 7",
    ],
    "title": "Question 3",
    "type": "FREE",
  },
  Object {
    "id": "4",
    "tags": Array [
      "Tag 1",
      "Tag 4",
      "Tag 5",
    ],
    "title": "Question 4",
    "type": "SC",
  },
]
`;

exports[`filterQuestions filters by tags 2`] = `
Array [
  Object {
    "id": "3",
    "tags": Array [
      "Tag 1",
      "Tag 6",
      "Tag 7",
    ],
    "title": "Question 3",
    "type": "FREE",
  },
]
`;

exports[`filterQuestions filters by tags 3`] = `Array []`;

exports[`filterQuestions filters by title 1`] = `
Array [
  Object {
    "id": "1",
    "tags": Array [
      "Tag 1",
      "Tag 2",
      "Tag 3",
    ],
    "title": "Question 1",
    "type": "SC",
  },
]
`;

exports[`filterQuestions filters by title 2`] = `Array []`;

exports[`filterQuestions filters by type 1`] = `
Array [
  Object {
    "id": "1",
    "tags": Array [
      "Tag 1",
      "Tag 2",
      "Tag 3",
    ],
    "title": "Question 1",
    "type": "SC",
  },
  Object {
    "id": "2",
    "tags": Array [
      "Tag 1",
      "Tag 4",
      "Tag 5",
    ],
    "title": "Question 2",
    "type": "MC",
  },
  Object {
    "id": "3",
    "tags": Array [
      "Tag 1",
      "Tag 6",
      "Tag 7",
    ],
    "title": "Question 3",
    "type": "FREE",
  },
  Object {
    "id": "4",
    "tags": Array [
      "Tag 1",
      "Tag 4",
      "Tag 5",
    ],
    "title": "Question 4",
    "type": "SC",
  },
]
`;

exports[`filterQuestions filters by type 2`] = `
Array [
  Object {
    "id": "1",
    "tags": Array [
      "Tag 1",
      "Tag 2",
      "Tag 3",
    ],
    "title": "Question 1",
    "type": "SC",
  },
  Object {
    "id": "2",
    "tags": Array [
      "Tag 1",
      "Tag 4",
      "Tag 5",
    ],
    "title": "Question 2",
    "type": "MC",
  },
  Object {
    "id": "3",
    "tags": Array [
      "Tag 1",
      "Tag 6",
      "Tag 7",
    ],
    "title": "Question 3",
    "type": "FREE",
  },
  Object {
    "id": "4",
    "tags": Array [
      "Tag 1",
      "Tag 4",
      "Tag 5",
    ],
    "title": "Question 4",
    "type": "SC",
  },
]
`;

exports[`filterQuestions returns all on empty filter 1`] = `
Array [
  Object {
    "id": "1",
    "tags": Array [
      "Tag 1",
      "Tag 2",
      "Tag 3",
    ],
    "title": "Question 1",
    "type": "SC",
  },
  Object {
    "id": "2",
    "tags": Array [
      "Tag 1",
      "Tag 4",
      "Tag 5",
    ],
    "title": "Question 2",
    "type": "MC",
  },
  Object {
    "id": "3",
    "tags": Array [
      "Tag 1",
      "Tag 6",
      "Tag 7",
    ],
    "title": "Question 3",
    "type": "FREE",
  },
  Object {
    "id": "4",
    "tags": Array [
      "Tag 1",
      "Tag 4",
      "Tag 5",
    ],
    "title": "Question 4",
    "type": "SC",
  },
]
`;
